Add max layer height to adaptive layers

vovodroid opened the pull request:
Add maximum layer height to adaptive layers dialog. Currently it's limited by extruder(s) max layer height, but in some situation it's desirable to limit it more, for example for better layers adhesion and strength.

Range is hardcoded 0.1-0.5, as I didn't succeed to get access to max_layer_height in Printer config, but I guess it's enough.

![image](https://user-images.githubusercontent.com/8691781/213760714-95e441b3-e536-44c8-848a-5688b15f3b28.png)

![image](https://user-images.githubusercontent.com/8691781/213760820-da3c9bdd-c68a-4497-9bf9-f0099ce4d86b.png)

vovodroid commented:
> you can manually adjust the slider for higher/lower heights, just click where the blue line are is.

It's for specific area only, not global limit

legend069 commented:
> > you can manually adjust the slider for higher/lower heights, just click where the blue line are is.
> 
> It's for specific area only, not global limit

I don't understand sorry.

It's rather easy to manually adjust what areas a user wants adjusted.
In the print settings tab (iirc that area ) 
It has the vales for adjusting min/max layers for adaptive layers.

vovodroid commented:
> It's rather easy to manually adjust what areas a user wants adjusted.

My idea is to limit layer height for the whole model.

vovodroid commented:
> I don't understand sorry.

Assuming we have 0.4 extrusion width. For good strength in Z direction it's desirable to keep layer height 0.2 or less.

legend069 commented:
> > It's rather easy to manually adjust what areas a user wants adjusted.
> 
> My idea is to limit layer height for the whole model.

sorry, i'm having trouble trying to understand.
is the new "feature" just moving the option for layer height limits?

or is it doing this ?
it can be useful if a user adjusts the slider applies the different layer height in a certain zone, then adjusts the slider again and applies a new max layer height. preventing getting high numbers like 0.12345679.
it would just result in being 0.15 adjustable layer ranges.


as it can already be done.
![image](https://user-images.githubusercontent.com/40685552/219938533-12d6bd97-744a-4312-8a20-39f973d4e0b1.png)

supermerill commented:
I reworked the inner working to use nozzle min/max and zstep, and also added a min.

vovodroid commented:
> I reworked the inner working to use nozzle min/max and zstep, and also added a min.

Excellent addition! I just didn't find how to access extruder config from this code.
Let's round min/max to 0.01mm or to Z-step (though I don't think Z-step parameter is necessary at all, 0.01mm is good enough).

vovodroid commented:
> I reworked the inner working to use nozzle min/max and zstep, and also added a min.

Shouldn't here https://github.com/supermerill/SuperSlicer/blob/nightly_dev/src/slic3r/GUI/GLCanvas3D.cpp#L282 
```
            min_height = std::min(min_height, float(extruders_min_height->get_abs_value(idx_extruder, nozzle_diameter->getFloat(idx_extruder))));
            max_height = std::max(max_height, float(extruders_max_height->get_abs_value(idx_extruder, nozzle_diameter->getFloat(idx_extruder))));
```
std::max and std::min be swapped? You would like to stay with any extruder range, wouldn't you?

vovodroid commented:
> I reworked the inner working to use nozzle min/max and zstep, and also added a min.

Shouldn't this https://github.com/supermerill/SuperSlicer/blob/nightly_dev/src/slic3r/GUI/GLCanvas3D.cpp#L277 code 
look like

```
        min_height =  0.f;
        max_height =std::numeric_limits<float>::max();
        assert(extruders_min_height->values.size() == extruders_max_height->values.size());
        assert(extruders_min_height->values.size() == nozzle_diameter->values.size());
        for (size_t idx_extruder = 0; idx_extruder < extruders_min_height->values.size(); ++idx_extruder) {
            min_height = std::max(min_height, float(extruders_min_height->get_abs_value(idx_extruder, nozzle_diameter->getFloat(idx_extruder))));
            max_height = std::min(max_height, float(extruders_max_height->get_abs_value(idx_extruder, nozzle_diameter->getFloat(idx_extruder))));
        }
```
this? (swap min and max)

We want to limit min and max heights stay within range of any extruder, don't we?

supermerill commented:
> We want to limit min and max heights stay within range of any extruder, don't we?

In an ideal world, yes.
But I don't know if I can have access to the sublist of the extruders that will be used on this object. And there may be another extruder with a min height higher than another max height.
So with this settings, it's a bit dangerous, but at least it won't prevent people to using the software.

vovodroid commented:
Merged
